File: packages/shared-auth/src/crypto.ts
import crypto from 'crypto';

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 16;
const SALT_LENGTH = 16;
const TAG_LENGTH = 16;

/**
 * Encrypts a string using AES-256-GCM.
 * The output format is: base64(salt + iv + auth_tag + ciphertext)
 */
export function encrypt(text: string, masterKey: string): string {
  if (!text) return text;
  if (!masterKey) throw new Error('Encryption key is required');

  // Generate salt and IV
  const salt = crypto.randomBytes(SALT_LENGTH);
  const iv = crypto.randomBytes(IV_LENGTH);

  // Derive a 32-byte key using PBKDF2
  const key = crypto.pbkdf2Sync(masterKey, salt, 100000, 32, 'sha256');

  // Create cipher
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv);

  // Encrypt the text
  let encrypted = cipher.update(text, 'utf8');
  encrypted = Buffer.concat([encrypted, cipher.final()]);

  // Extract the auth tag
  const authTag = cipher.getAuthTag();

  // Concatenate all parts and base64 encode
  // Format: [salt (16)] [iv (16)] [tag (16)] [encrypted...]
  const payload = Buffer.concat([salt, iv, authTag, encrypted]);
  return payload.toString('base64');
}

/**
 * Decrypts a string previously encrypted with encrypt().
 */
export function decrypt(encryptedData: string, masterKey: string): string {
  if (!encryptedData) return encryptedData;
  if (!masterKey) throw new Error('Decryption key is required');

  try {
    const payload = Buffer.from(encryptedData, 'base64');

    // Extract parts
    const salt = payload.subarray(0, SALT_LENGTH);
    const iv = payload.subarray(SALT_LENGTH, SALT_LENGTH + IV_LENGTH);
    const authTag = payload.subarray(SALT_LENGTH + IV_LENGTH, SALT_LENGTH + IV_LENGTH + TAG_LENGTH);
    const encryptedText = payload.subarray(SALT_LENGTH + IV_LENGTH + TAG_LENGTH);

    // Derive the key
    const key = crypto.pbkdf2Sync(masterKey, salt, 100000, 32, 'sha256');

    // Create decipher
    const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
    decipher.setAuthTag(authTag);

    // Decrypt the text
    let decrypted = decipher.update(encryptedText);
    decrypted = Buffer.concat([decrypted, decipher.final()]);

    return decrypted.toString('utf8');
  } catch (error) {
    throw new Error('Failed to decrypt data. Invalid key or corrupted data.');
  }
}

File: services/orchestrator/crypto.py
import base64
import os
import hashlib
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend

SALT_LENGTH = 16
IV_LENGTH = 16
TAG_LENGTH = 16
ITERATIONS = 100000
KEY_LENGTH = 32

def get_master_key() -> str:
    key = os.getenv("ENCRYPTION_KEY")
    if not key:
        raise RuntimeError(
            "ENCRYPTION_KEY environment variable is required and was not set. "
            "Generate one with: python -c \"import secrets; print(secrets.token_hex(32))\""
        )
    return key

def decrypt(encrypted_data_b64: str, master_key: str = None) -> str:
    """
    Decrypts a base64 encoded string that was encrypted by the Node.js shared-auth/crypto.ts
    Format: base64( [salt (16)] [iv (16)] [tag (16)] [ciphertext] )
    """
    if not encrypted_data_b64:
        return encrypted_data_b64
        
    master_key = master_key or get_master_key()
    
    try:
        # Check if the string is actually JSON (plain text config) for backward compatibility
        if encrypted_data_b64.strip().startswith('{') or encrypted_data_b64.strip().startswith('['):
            return encrypted_data_b64
            
        payload = base64.b64decode(encrypted_data_b64)
        
        salt = payload[:SALT_LENGTH]
        iv = payload[SALT_LENGTH : SALT_LENGTH + IV_LENGTH]
        auth_tag = payload[SALT_LENGTH + IV_LENGTH : SALT_LENGTH + IV_LENGTH + TAG_LENGTH]
        ciphertext = payload[SALT_LENGTH + IV_LENGTH + TAG_LENGTH:]
        
        # Derive key using PBKDF2
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=KEY_LENGTH,
            salt=salt,
            iterations=ITERATIONS,
            backend=default_backend()
        )
        key = kdf.derive(master_key.encode('utf-8'))
        
        # Python cryptography AESGCM expects the ciphertext + auth_tag together
        aesgcm = AESGCM(key)
        data = ciphertext + auth_tag
        decrypted = aesgcm.decrypt(iv, data, None)
        
        return decrypted.decode('utf-8')
    except Exception as e:
        print(f"Failed to decrypt provider config: {e}")
        # If decryption fails, maybe it was just plaintext anyway
        return encrypted_data_b64

def encrypt(text: str, master_key: str = None) -> str:
    """
    Encrypts a string, matching the format expected by the Node.js shared-auth/crypto.ts
    """
    if not text:
        return text
        
    master_key = master_key or get_master_key()
    
    salt = os.urandom(SALT_LENGTH)
    iv = os.urandom(IV_LENGTH)
    
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=KEY_LENGTH,
        salt=salt,
        iterations=ITERATIONS,
        backend=default_backend()
    )
    key = kdf.derive(master_key.encode('utf-8'))
    
    aesgcm = AESGCM(key)
    # Encrypt returns ciphertext + auth_tag
    encrypted_data = aesgcm.encrypt(iv, text.encode('utf-8'), None)
    
    ciphertext = encrypted_data[:-TAG_LENGTH]
    auth_tag = encrypted_data[-TAG_LENGTH:]
    
    payload = salt + iv + auth_tag + ciphertext
    return base64.b64encode(payload).decode('utf-8')
